src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcClient.cs

```diff
@@ -3,89 +3,23 @@
-        /// </summary>
-        /// <param name="processId">The PID of the dotnet process to get the transport for</param>
-        /// <returns>A System.IO.Stream wrapper around the transport</returns>
-        private static Stream GetTransport(int processId)
-        {
-            try 
-            {
-                var process = Process.GetProcessById(processId);
-            }
-            catch (System.ArgumentException)
-            {
-                throw new ServerNotAvailableException($"Process {processId} is not running.");
-            }
-            catch (System.InvalidOperationException)
-            {
-                throw new ServerNotAvailableException($"Process {processId} seems to be elevated.");
-            }
- 
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-            {
-                string pipeName = $"dotnet-diagnostic-{processId}";
-                var namedPipe = new NamedPipeClientStream(
-                    ".", pipeName, PipeDirection.InOut, PipeOptions.None, TokenImpersonationLevel.Impersonation);
-                namedPipe.Connect((int)ConnectTimeoutMilliseconds);
-                return namedPipe;
-            }
-            else
-            {
-                string ipcPort;
-                try
-                {
-                    ipcPort = Directory.GetFiles(IpcRootPath, $"dotnet-diagnostic-{processId}-*-socket") // Try best match.
-                                .OrderByDescending(f => new FileInfo(f).LastWriteTime)
-                                .FirstOrDefault();
-                    if (ipcPort == null)
-                    {
-                        throw new ServerNotAvailableException($"Process {processId} not running compatible .NET Core runtime.");
-                    }
-                }
-                catch (InvalidOperationException)
-                {
-                    throw new ServerNotAvailableException($"Process {processId} not running compatible .NET Core runtime.");
-                }
-                string path = Path.Combine(IpcRootPath, ipcPort);
-                var remoteEP = CreateUnixDomainSocketEndPoint(path);
-
-                var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
-                socket.Connect(remoteEP);
-                return new NetworkStream(socket, ownsSocket: true);
-            }
-        }
-
         /// <summary>
         /// Checks that the client is able to communicate with target process over diagnostic transport.
         /// </summary>
         /// <returns>
         /// True if client is able to communicate with target process; otherwise, false.
         /// </returns>
-        public static bool CheckTransport(int processId)
+        public static bool CheckTransport(IIpcEndpoint endpoint)
         {
             try
             {
-                using var stream = GetTransport(processId);
+                using var stream = endpoint.Connect();
```

Comment:
If you use this method with a reverse server won't your implementation of `ServerIpcEndpoint.Connect` cause the underlying stream value to become `null`?  This makes the  check mutating.

Reply:
The original intent of the `CheckTransport` method was to allow clients to check if the client is able to communicate to a target process via the diagnostics transport. Something along the lines of:

```
var client = new DiagnosticsClient(pid);
var hasTransport = client.CheckTransport()
```

We specifically use this is dotnet-monitor to check if process 1 in a Docker container is actually a .NET process that has the diagnostics transport (otherwise, we want to fall back to the default process selection algorithm).

After this refactor, the above behavior remains unchanged for the PID endpoint. The check was always mutating previous to the refactor, because it would always hand out a new NamedPipeClientStream/NetworkStream (but it's not really mutating because it didn't cache the stream).

The refactor introduces the server based endpoint, which does store the NamedPipeSeverStream/NetworkStream. When it is retrieved and acted upon or closed, the target runtime instance should be reconnecting to the server, which will then fill in the stream once again for subsequent usage. Any kind of usage of the stream in the reversed server case needs to be mutating so that the runtime instance will reconnect with a new stream for additional requests.

Maybe I can refactor the CheckTransport concept a bit for the different endpoint types. Possibly:
- add a Check* method to the IIpcEndpoint interface.
- Have PidIpcClient check if the pipe/socket exists (without connecting to it).
- Have ServerIpcClient check if the stream is not null.
This should preserve the intended check and not consume any connections in either case.